poly/promise/promise.go

```diff
@@ -0,0 +1,1355 @@
+ *     clients.
+ *
+ * 11) Clients can then request the signatures from the insurers to make sure
+ *     the Promise is indeed certified
+ *
+ *      i) This prevents a malicious promiser from simply leaving out valid
+ *         BlameProofs and only sending good signatures to the clients.
+ *
+ * 12) The promiser can perform some work for clients.
+ *
+ * 13) If the promiser is unresponsive for too long, a client can inform the
+ *     insurers of the Promise.
+ *
+ * 14) The insurers can then check if the promiser is indeed unresponsive.
+ *
+ * 15) If so, an insurer reveal its share and send it to the client.
+ *
+ * 16) The client repeats steps 13-15 until enough shares are recovered to
+ *     reconstruct the secret.
+ *
+ * 17) The client can then reconstruct the secret and take over for the promiser
+ *
+ *
+ *
+ * Key Terms:
+ *   promiser = the server making a Promise
+ *   client   = recipients of a Promise who are trusting the promiser
+ *   insurer  = servers who store secret shares of a promise. Such servers help
+ *              during secret reconstruction.
+ *
+ *   Users of this code = programmers wishing to use this code in programs
+*/
+
+// Used mostly in marshalling code, this is the size of a uint32
+var uint32Size int = binary.Size(uint32(0))
+
+// This is the protocol name used by crypto/proof verifiers and provers.
+var protocolName string = "Promise Protocol"
+
+// These are messages used for signatures
+var sigMsg []byte = []byte("Promise Signature")
+var sigBlameMsg []byte = []byte("Promise Blame Signature")
+
+/* Promise structs are mechanisms by which a server can promise other servers
+ * that an abstract.Secret will be availble even if the secret's owner goes
+ * down. The secret to be promised will be sharded into shared secrets that can
+ * be combined using Lagrange Interpolation to produce the original secret.
+ * Other servers will act as insurers maintaining a share. If a client ever
+ * needs the secret to be reconstructed, it can contact the insurers to regain
+ * the shares and reconstruct the secret.
+ *
+ * Note to Developers:
+ *       The insurers and secrets arrays should remain synchronized. In other
+ *       words, insurers[i] and secrets[i] should both refer to the same server.
+ *
+ * Note to users of this code:
+ *
+ *   Here is a list of methods that should be called by each type of server:
+ *
+ * - Promisers
+ *   * ConstructPromise
+ *
+ * - Insurers
+ *   * VerifyPromise
+ *   * VerifyShare
+ *   * Sign
+ *   * RevealShare
+ *   * Blame
+ *
+ * - Clients
+ *   * VerifyPromise
+ *   * VerifyRevealedShare
+ *
+ * - All
+ *   * UnmarshalInit
+ *   * VerifySignature
+ *   * VerifyBlame
+ */
+type Promise struct {
+
+	// The id of the promise used to differentiate it from others
+	// The id is the short term public key of the private key being promised
+	id string
+
+	// The cryptographic suite to use for the shared secrets
+	shareSuite abstract.Suite
```

Comment:
Can shareSuite and keySuite ever be different?  I don't think so, because the secret-shared Secret needs to be usable and thus in an appropriate modulus for both secret-sharing and DH key exchange with the long-term keys.  I think you could probably just have a single 'suite' config variable here.